### PR TITLE
Adding start up retries.

### PR DESCRIPTION
**Pull request by Tingting-He-ODC**

### CODE REVIEW

File: src/main/java/com/addthis/meshy/Backoff.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
+package com.addthis.meshy;
```
Comment (tedpearson) [inline src/main/java/com/addthis/meshy/Backoff.java]: Missing license header

File: src/main/java/com/addthis/meshy/MeshyServer.java [hunk 2]
```diff
@@ -77,6 +78,10 @@
     private static final boolean allowPeerLocal = Parameter.boolValue("meshy.peer.local", true);
     private static final int autoMeshTimeout = Parameter.intValue("meshy.autoMeshTimeout", 60000);
 
+    private static final int meshyUnitDelay = Parameter.intValue("mss.mesh.unitDelay", 900);
+    private static final int meshyMaxDelayUnit = Parameter.intValue("mss.mesh.maxDelayUnit", 3);
+    private static final int meshyMaxRetries = Parameter.intValue("mss.mesh.maxRetries", 5);
```
Comment (tedpearson) [inline src/main/java/com/addthis/meshy/MeshyServer.java]: The parameter names and the variable names here do not make it obvious at all that they are for connection retries. We should make it more obvious with better names.

File: src/main/java/com/addthis/meshy/MeshyServer.java [hunk 3]
```diff
@@ -384,6 +389,25 @@ public void connectPeer(InetSocketAddress address) {
...
+    public void connectPeers(List<InetSocketAddress> addresses) {
+        connectPeers(addresses, meshyUnitDelay, meshyMaxDelayUnit, meshyMaxRetries);
+    }
+
+    public void connectPeers(List<InetSocketAddress> addresses, int unitDelay, int maxDelayUnit, int maxRetries) {
+        Backoff backoff = new Backoff(unitDelay, maxDelayUnit);
+        int retries=0;
+        while(getChannelCount()==0 && retries++<maxRetries) {
+            if (retries > 1) {
+                log.info("gagagaga mss will retry in: " + backoff.calcDelay() + " milliseconds.");
```
Comment (tedpearson) [inline src/main/java/com/addthis/meshy/MeshyServer.java]: Let's make sure we remove the gagagaga before merging; also say what is being retried (the connection to seeds)

File: src/main/java/com/addthis/meshy/MeshyServer.java [hunk 4]
```diff
@@ -77,6 +78,10 @@
     private static final boolean allowPeerLocal = Parameter.boolValue("meshy.peer.local", true);
     private static final int autoMeshTimeout = Parameter.intValue("meshy.autoMeshTimeout", 60000);
 
+    private static final int retryPeersUnitDelay = Parameter.intValue("mss.mesh.retry.peers.unitDelay", 900);
+    private static final int retryPeersMaxDelayUnit = Parameter.intValue("mss.mesh.retry.peers.maxDelayUnit", 3);
+    private static final int retryPeersMaxRetries = Parameter.intValue("mss.mesh.retry.peers.maxRetries", 5);
```
Comment (tedpearson) [inline src/main/java/com/addthis/meshy/MeshyServer.java]: How about `mss.mesh.connection.retry.maxDelayUnit`, `mss.mesh.connection.retry.unitDelay`, `mss.mesh.connection.retry.maxRetries`?

Comment (tedpearson) [reply src/main/java/com/addthis/meshy/MeshyServer.java]: Feel free to disagree, not sure if mine or yours is better.

**Comment by Tingting-He-ODC** [comment]
Not only startup retries, the retry connection to peers will happen all the time, more frequent during startup time. Retry one time per minute afterwards. If already connected to all seeds, the operation will return a null future:
`INFO [AutoConnectToPeers] 2019-06-03T19:32:02,307 MeshyServer.java (line 382) MeshyServer{serverPort=7123, serverUuid=cloudy-hydra-test-phx-63-1-d35290-7123, channelCount=6, peeredCount=6} peer connect returned null future to cloudy-mqmaster-test-phx-63-1.wh1phx1.oracledatacloud.com/10.196.1.244:7123
INFO [AutoConnectToPeers] 2019-06-03T19:32:02,307 MeshyServer.java (line 382) MeshyServer{serverPort=7123, serverUuid=cloudy-hydra-test-phx-63-1-d35290-7123, channelCount=6, peeredCount=6} peer connect returned null future to cloudy-hydra-test-phx-63-1.wh1phx1.oracledatacloud.com/10.196.1.245:7123
INFO [AutoConnectToPeers] 2019-06-03T19:32:02,307 AutoConnectToPeersTask.java (line 56) mss will retry connection to seeds in: 28800 milliseconds.`

File: src/main/java/com/addthis/meshy/AutoConnectToPeersTask.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        this.maxDelayUnit = maxDelayUnit;
+    }
+
+    @Override public void run() {
+        try {
+            Backoff backoff = new Backoff(unitDelay, maxDelayUnit);
+            while (true) {
+                log.info("mss will retry connection to seeds in: " + backoff.calcDelay() + " milliseconds.");
+                backoff.backoff();
+                backoff.inc();
```
Comment (tedpearson) [inline src/main/java/com/addthis/meshy/AutoConnectToPeersTask.java]: The backoff stuff should only be used to back off from error, rate limits, or things of the sort. Do we actually need a backoff for this task?

File: src/main/java/com/addthis/meshy/MeshyServer.java [hunk 6]
```diff
@@ -384,6 +388,14 @@ public void connectPeer(InetSocketAddress address) {
         }
     }
 
+    public void connectPeers(List<InetSocketAddress> addresses) {
+        connectPeers(addresses, retryPeersUnitDelay, retryPeersMaxDelayUnit);
+    }
+
+    public void connectPeers(List<InetSocketAddress> addresses, int unitDelay, int maxDelayUnit) {
```
Comment (tedpearson) [inline src/main/java/com/addthis/meshy/MeshyServer.java]: I think we should move the startAutoConnectToPeers() call into the MeshyServer constructor:
`if (automesh) {startAutoMesh} else {startAutoConnectToPeers}`

I think if we do this, we can undo all the changes to meshy/hydra/mqmaster (except for upgrading meshy version) and get rid of both connectPeers methods.

What do you think?

**Pull request merged**